// @sentry/browser-utils/build/esm/index.js
export { addClsInstrumentationHandler, addInpInstrumentationHandler, addLcpInstrumentationHandler, addPerformanceInstrumentationHandler, addTtfbInstrumentationHandler } from './metrics/instrument.js';
export { addPerformanceEntries, addWebVitalsToSpan, startTrackingInteractions, startTrackingLongAnimationFrames, startTrackingLongTasks, startTrackingWebVitals } from './metrics/browserMetrics.js';
export { elementTimingIntegration, startTrackingElementTiming } from './metrics/elementTiming.js';
export { extractNetworkProtocol } from './metrics/utils.js';
export { trackClsAsSpan, trackInpAsSpan, trackLcpAsSpan } from './metrics/webVitalSpans.js';
export { whenIdleOrHidden } from './metrics/web-vitals/lib/whenIdleOrHidden.js';
export { addClickKeypressInstrumentationHandler } from './instrument/dom.js';
export { addHistoryInstrumentationHandler } from './instrument/history.js';
export { clearCachedImplementation, fetch, getNativeImplementation, setTimeout } from './getNativeImplementation.js';
export { SENTRY_XHR_DATA_KEY, addXhrInstrumentationHandler } from './instrument/xhr.js';
export { getBodyString, getFetchRequestArgBody, parseXhrResponseHeaders, serializeFormData } from './networkUtils.js';
export { resourceTimingToSpanAttributes } from './metrics/resourceTiming.js';
export { htmlTreeAsString } from './htmlTreeAsString.js';
export { isElement } from './is.js';
export { getAbsoluteUrl } from './instrument/location.js';
export { registerInpInteractionListener, startTrackingINP } from './metrics/inp.js';
