// @sentry/react/build/esm/reactrouter-compat-utils/instrumentation.js
import { WINDOW, browserTracingIntegration, startBrowserTracingPageLoadSpan, startBrowserTracingNavigationSpan } from '@sentry/browser';
import { SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN, SEMANTIC_ATTRIBUTE_SENTRY_OP, SEMANTIC_ATTRIBUTE_SENTRY_SOURCE, debug, addNonEnumerableProperty, spanToJSON, getCurrentScope, getClient } from '@sentry/core';
import * as React from 'react';
import { DEBUG_BUILD } from '../debug-build.js';
import { hoistNonReactStatics } from '../hoist-non-react-statics.js';
import { checkRouteForAsyncHandler } from './lazy-routes.js';
import { getActiveRootSpan, setNavigationContext, clearNavigationContext, resolveRouteNameAndSource, transactionNameHasWildcard, initializeRouterUtils } from './utils.js';
import { URL_TEMPLATE } from '@sentry/conventions/attributes';

let _useEffect;
let _useLocation;
let _useNavigationType;
let _createRoutesFromChildren;
let _matchRoutes;
let _enableAsyncRouteHandlers = false;
let _lazyRouteTimeout = 3e3;
let _lazyRouteManifest;
let _basename = "";
const CLIENTS_WITH_INSTRUMENT_NAVIGATION = /* @__PURE__ */ new WeakSet();
const useIsomorphicLayoutEffect = WINDOW?.document ? React.useLayoutEffect : React.useEffect;
const activeNavigationSpans = /* @__PURE__ */ new WeakMap();
const allRoutes = /* @__PURE__ */ new Set();
const pendingLazyRouteLoads = /* @__PURE__ */ new WeakMap();
const deferredLazyRouteResolvers = /* @__PURE__ */ new WeakMap();
function scheduleCallback(callback) {
  if (WINDOW?.requestAnimationFrame) {
    return WINDOW.requestAnimationFrame(callback);
  }
  return setTimeout(callback, 0);
}
function cancelScheduledCallback(id) {
  if (WINDOW?.cancelAnimationFrame) {
    WINDOW.cancelAnimationFrame(id);
  } else {
    clearTimeout(id);
  }
}
function computeLocationKey(location) {
  return `${location.pathname}${location.search || ""}${location.hash || ""}`;
}
function isParameterizedRoute(routeName) {
  return routeName.includes(":") || routeName.includes("*");
}
function shouldSkipNavigation(trackedNav, locationKey, proposedName, spanHasEnded) {
  if (!trackedNav) {
    return { skip: false, shouldUpdate: false };
  }
  const isDuplicate = trackedNav.locationKey === locationKey && (trackedNav.isPlaceholder || !spanHasEnded);
  if (isDuplicate) {
    const currentHasWildcard = !!trackedNav.routeName && transactionNameHasWildcard(trackedNav.routeName);
    const proposedHasWildcard = transactionNameHasWildcard(proposedName);
    const currentIsParameterized = !!trackedNav.routeName && isParameterizedRoute(trackedNav.routeName);
    const proposedIsParameterized = isParameterizedRoute(proposedName);
    const isWildcardUpgrade = currentHasWildcard && !proposedHasWildcard;
    const isRawToParameterized = !currentIsParameterized && proposedIsParameterized;
    const isMoreSpecific = proposedName !== trackedNav.routeName && proposedName.length > (trackedNav.routeName?.length || 0) && !proposedHasWildcard;
    const shouldUpdate = !!(trackedNav.routeName && (isWildcardUpgrade || isRawToParameterized || isMoreSpecific));
    return { skip: true, shouldUpdate };
  }
  return { skip: false, shouldUpdate: false };
}
function addResolvedRoutesToParent(resolvedRoutes, parentRoute) {
  const existingChildren = parentRoute.children || [];
  const newRoutes = resolvedRoutes.filter(
    (newRoute) => !existingChildren.some(
      (existing) => existing === newRoute || newRoute.path && existing.path === newRoute.path || newRoute.id && existing.id === newRoute.id
    )
  );
  if (newRoutes.length > 0) {
    parentRoute.children = [...existingChildren, ...newRoutes];
  }
}
function trackLazyRouteLoad(span, promise) {
  let promises = pendingLazyRouteLoads.get(span);
  if (!promises) {
    promises = /* @__PURE__ */ new Set();
    pendingLazyRouteLoads.set(span, promises);
  }
  promises.add(promise);
  promise.finally(() => {
    const currentPromises = pendingLazyRouteLoads.get(span);
    if (currentPromises) {
      currentPromises.delete(promise);
    }
  });
}
function createDeferredLazyRoutePromise(span) {
  const deferredPromise = new Promise((resolve) => {
    deferredLazyRouteResolvers.set(span, resolve);
  });
  trackLazyRouteLoad(span, deferredPromise);
}
function resolveDeferredLazyRoutePromise(span) {
  const resolver = deferredLazyRouteResolvers.get(span);
  if (resolver) {
    resolver();
    deferredLazyRouteResolvers.delete(span);
    if (span.__sentry_may_have_lazy_routes__) {
      span.__sentry_may_have_lazy_routes__ = false;
    }
  }
}
function processResolvedRoutes(resolvedRoutes, parentRoute, currentLocation = null, capturedSpan) {
  resolvedRoutes.forEach((child) => {
    allRoutes.add(child);
    if (_enableAsyncRouteHandlers) {
      checkRouteForAsyncHandler(child, processResolvedRoutes);
    }
  });
  if (parentRoute) {
    addResolvedRoutesToParent(resolvedRoutes, parentRoute);
  }
  const targetSpan = capturedSpan ?? getActiveRootSpan();
  if (targetSpan) {
    const spanJson = spanToJSON(targetSpan);
    if (spanJson.timestamp) {
      DEBUG_BUILD && debug.warn("[React Router] Lazy handler resolved after span ended - skipping update");
      return;
    }
    const spanOp = spanJson.op;
    let location = currentLocation;
    if (!location && !capturedSpan) {
      if (typeof WINDOW !== "undefined") {
        const globalLocation = WINDOW.location;
        if (globalLocation?.pathname) {
          location = { pathname: globalLocation.pathname };
        }
      }
    }
    if (location) {
      if (spanOp === "pageload") {
        updatePageloadTransaction({
          activeRootSpan: targetSpan,
          location: { pathname: location.pathname },
          routes: Array.from(allRoutes),
          allRoutes: Array.from(allRoutes)
        });
      } else if (spanOp === "navigation") {
        updateNavigationSpan(targetSpan, location, Array.from(allRoutes), false, _matchRoutes);
      }
    }
  }
}
function updateNavigationSpan(activeRootSpan, location, allRoutes2, forceUpdate = false, matchRoutes) {
  const spanJson = spanToJSON(activeRootSpan);
  const currentName = spanJson.description;
  const hasBeenNamed = activeRootSpan?.__sentry_navigation_name_set__;
  const currentNameHasWildcard = currentName && transactionNameHasWildcard(currentName);
  const shouldUpdate = !hasBeenNamed || forceUpdate || currentNameHasWildcard;
  if (shouldUpdate && !spanJson.timestamp) {
    const currentBranches = matchRoutes(allRoutes2, location);
    const [name, source] = resolveRouteNameAndSource(
      location,
      allRoutes2,
      allRoutes2,
      currentBranches || [],
      _basename,
      _lazyRouteManifest,
      _enableAsyncRouteHandlers
    );
    const currentSource = spanJson.data?.[SEMANTIC_ATTRIBUTE_SENTRY_SOURCE];
    const isImprovement = name && (!currentName || // No current name - always set
    !hasBeenNamed && (currentSource !== "route" || source === "route") || // Not finalized - allow unless downgrading route→url
    currentSource !== "route" && source === "route" || // URL → route upgrade
    currentSource === "route" && source === "route" && currentNameHasWildcard);
    if (isImprovement) {
      activeRootSpan.updateName(name);
      activeRootSpan.setAttribute(SEMANTIC_ATTRIBUTE_SENTRY_SOURCE, source);
      if (source === "route") {
        activeRootSpan.setAttribute(URL_TEMPLATE, name);
      }
      if (!transactionNameHasWildcard(name) && source === "route") {
        addNonEnumerableProperty(
          activeRootSpan,
          "__sentry_navigation_name_set__",
          true
        );
      }
    }
  }
}
function setupRouterSubscription(router, routes, version, basename, activeRootSpan) {
  let isInitialPageloadComplete = false;
  let hasSeenPageloadSpan = !!activeRootSpan && spanToJSON(activeRootSpan).op === "pageload";
  let hasSeenPopAfterPageload = false;
  let scheduledNavigationHandler = null;
  let lastHandledPathname = null;
  router.subscribe((state) => {
    if (!isInitialPageloadComplete) {
      const currentRootSpan = getActiveRootSpan();
      const isCurrentlyInPageload = currentRootSpan && spanToJSON(currentRootSpan).op === "pageload";
      if (isCurrentlyInPageload) {
        hasSeenPageloadSpan = true;
      } else if (hasSeenPageloadSpan) {
        if (state.historyAction === "POP" && !hasSeenPopAfterPageload) {
          hasSeenPopAfterPageload = true;
        } else {
          isInitialPageloadComplete = true;
        }
      }
    }
    const shouldHandleNavigation = state.historyAction === "PUSH" || state.historyAction === "POP" && isInitialPageloadComplete;
    if (shouldHandleNavigation) {
      const currentLocationKey = computeLocationKey(state.location);
      const navigationHandler = () => {
        if (lastHandledPathname === currentLocationKey) {
          return;
        }
        lastHandledPathname = currentLocationKey;
        scheduledNavigationHandler = null;
        handleNavigation({
          location: state.location,
          routes,
          navigationType: state.historyAction,
          version,
          basename,
          allRoutes: Array.from(allRoutes)
        });
      };
      if (state.navigation.state !== "idle") {
        if (lastHandledPathname !== currentLocationKey) {
          lastHandledPathname = null;
        }
        if (scheduledNavigationHandler !== null) {
          cancelScheduledCallback(scheduledNavigationHandler);
        }
        scheduledNavigationHandler = scheduleCallback(navigationHandler);
      } else {
        if (scheduledNavigationHandler !== null) {
          cancelScheduledCallback(scheduledNavigationHandler);
          scheduledNavigationHandler = null;
        }
        navigationHandler();
      }
    }
  });
}
function createV6CompatibleWrapCreateBrowserRouter(createRouterFunction, version) {
  if (!_useEffect || !_useLocation || !_useNavigationType || !_matchRoutes) {
    DEBUG_BUILD && debug.warn(
      `reactRouter${version ? `V${version}` : ""}Instrumentation was unable to wrap the \`createRouter\` function because of one or more missing parameters.`
    );
    return createRouterFunction;
  }
  return function(routes, opts) {
    addRoutesToAllRoutes(routes);
    if (_enableAsyncRouteHandlers) {
      for (const route of routes) {
        checkRouteForAsyncHandler(route, processResolvedRoutes);
      }
    }
    const activeRootSpan = getActiveRootSpan();
    const hasPatchRoutesOnNavigation = opts && "patchRoutesOnNavigation" in opts && typeof opts.patchRoutesOnNavigation === "function";
    if (hasPatchRoutesOnNavigation && activeRootSpan) {
      addNonEnumerableProperty(
        activeRootSpan,
        "__sentry_may_have_lazy_routes__",
        true
      );
      createDeferredLazyRoutePromise(activeRootSpan);
    }
    const wrappedOpts = wrapPatchRoutesOnNavigation(opts, false, activeRootSpan);
    const router = createRouterFunction(routes, wrappedOpts);
    const basename = opts?.basename;
    if (router.state.historyAction === "POP" && activeRootSpan) {
      updatePageloadTransaction({
        activeRootSpan,
        location: router.state.location,
        routes,
        basename,
        allRoutes: Array.from(allRoutes)
      });
    }
    _basename = basename || "";
    setupRouterSubscription(router, routes, version, basename, activeRootSpan);
    return router;
  };
}
function createV6CompatibleWrapCreateMemoryRouter(createRouterFunction, version) {
  if (!_useEffect || !_useLocation || !_useNavigationType || !_matchRoutes) {
    DEBUG_BUILD && debug.warn(
      `reactRouter${version ? `V${version}` : ""}Instrumentation was unable to wrap the \`createMemoryRouter\` function because of one or more missing parameters.`
    );
    return createRouterFunction;
  }
  return function(routes, opts) {
    addRoutesToAllRoutes(routes);
    if (_enableAsyncRouteHandlers) {
      for (const route of routes) {
        checkRouteForAsyncHandler(route, processResolvedRoutes);
      }
    }
    const memoryActiveRootSpanEarly = getActiveRootSpan();
    const hasPatchRoutesOnNavigation = opts && "patchRoutesOnNavigation" in opts && typeof opts.patchRoutesOnNavigation === "function";
    if (hasPatchRoutesOnNavigation && memoryActiveRootSpanEarly) {
      addNonEnumerableProperty(
        memoryActiveRootSpanEarly,
        "__sentry_may_have_lazy_routes__",
        true
      );
      createDeferredLazyRoutePromise(memoryActiveRootSpanEarly);
    }
    const wrappedOpts = wrapPatchRoutesOnNavigation(opts, true, memoryActiveRootSpanEarly);
    const router = createRouterFunction(routes, wrappedOpts);
    const basename = opts?.basename;
    let initialEntry = void 0;
    const initialEntries = opts?.initialEntries;
    const initialIndex = opts?.initialIndex;
    const hasOnlyOneInitialEntry = initialEntries?.length === 1;
    const hasIndexedEntry = initialIndex !== void 0 && initialEntries?.[initialIndex];
    initialEntry = hasOnlyOneInitialEntry ? initialEntries[0] : hasIndexedEntry ? initialEntries[initialIndex] : void 0;
    const location = initialEntry ? typeof initialEntry === "string" ? { pathname: initialEntry } : initialEntry : router.state.location;
    const memoryActiveRootSpan = getActiveRootSpan();
    if (router.state.historyAction === "POP" && memoryActiveRootSpan) {
      updatePageloadTransaction({
        activeRootSpan: memoryActiveRootSpan,
        location,
        routes,
        basename,
        allRoutes: Array.from(allRoutes)
      });
    }
    _basename = basename || "";
    setupRouterSubscription(router, routes, version, basename, memoryActiveRootSpan);
    return router;
  };
}
function createReactRouterV6CompatibleTracingIntegration(options, version) {
  const integration = browserTracingIntegration({ ...options, instrumentPageLoad: false, instrumentNavigation: false });
  const {
    useEffect,
    useLocation,
    useNavigationType,
    createRoutesFromChildren,
    matchRoutes,
    stripBasename,
    enableAsyncRouteHandlers = false,
    instrumentPageLoad = true,
    instrumentNavigation = true,
    lazyRouteTimeout,
    lazyRouteManifest
  } = options;
  return {
    ...integration,
    setup(client) {
      integration.setup(client);
      const finalTimeout = options.finalTimeout ?? 3e4;
      const defaultMaxWait = (options.idleTimeout ?? 1e3) * 3;
      const configuredMaxWait = lazyRouteTimeout ?? defaultMaxWait;
      if (configuredMaxWait === Infinity) {
        _lazyRouteTimeout = finalTimeout;
        DEBUG_BUILD && debug.log(
          "[React Router] lazyRouteTimeout set to Infinity, capping at finalTimeout:",
          finalTimeout,
          "ms to prevent indefinite hangs"
        );
      } else if (Number.isNaN(configuredMaxWait)) {
        DEBUG_BUILD && debug.warn("[React Router] lazyRouteTimeout must be a number, falling back to default:", defaultMaxWait);
        _lazyRouteTimeout = defaultMaxWait;
      } else if (configuredMaxWait < 0) {
        DEBUG_BUILD && debug.warn(
          "[React Router] lazyRouteTimeout must be non-negative or Infinity, got:",
          configuredMaxWait,
          "falling back to:",
          defaultMaxWait
        );
        _lazyRouteTimeout = defaultMaxWait;
      } else {
        _lazyRouteTimeout = configuredMaxWait;
      }
      _useEffect = useEffect;
      _useLocation = useLocation;
      _useNavigationType = useNavigationType;
      _matchRoutes = matchRoutes;
      _createRoutesFromChildren = createRoutesFromChildren;
      _enableAsyncRouteHandlers = enableAsyncRouteHandlers;
      _lazyRouteManifest = lazyRouteManifest;
      initializeRouterUtils(matchRoutes, stripBasename || false);
    },
    afterAllSetup(client) {
      integration.afterAllSetup(client);
      const initPathName = WINDOW.location?.pathname;
      if (instrumentPageLoad && initPathName) {
        startBrowserTracingPageLoadSpan(client, {
          name: initPathName,
          attributes: {
            [SEMANTIC_ATTRIBUTE_SENTRY_SOURCE]: "url",
            [SEMANTIC_ATTRIBUTE_SENTRY_OP]: "pageload",
            [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: `auto.pageload.react.reactrouter${version ? `_v${version}` : ""}`
          }
        });
      }
      if (instrumentNavigation) {
        CLIENTS_WITH_INSTRUMENT_NAVIGATION.add(client);
      }
    }
  };
}
function createV6CompatibleWrapUseRoutes(origUseRoutes, version) {
  if (!_useEffect || !_useLocation || !_useNavigationType || !_matchRoutes) {
    DEBUG_BUILD && debug.warn(
      "reactRouterV6Instrumentation was unable to wrap `useRoutes` because of one or more missing parameters."
    );
    return origUseRoutes;
  }
  const SentryRoutes = (props) => {
    const isMountRenderPass = React.useRef(true);
    const { routes, locationArg } = props;
    const Routes = origUseRoutes(routes, locationArg);
    const location = _useLocation();
    const navigationType = _useNavigationType();
    const stableLocationParam = typeof locationArg === "string" || locationArg?.pathname ? locationArg : location;
    useIsomorphicLayoutEffect(() => {
      const normalizedLocation = typeof stableLocationParam === "string" ? { pathname: stableLocationParam } : stableLocationParam;
      if (isMountRenderPass.current) {
        addRoutesToAllRoutes(routes);
        updatePageloadTransaction({
          activeRootSpan: getActiveRootSpan(),
          location: normalizedLocation,
          routes,
          allRoutes: Array.from(allRoutes)
        });
        isMountRenderPass.current = false;
      } else {
        handleNavigation({
          location: normalizedLocation,
          routes,
          navigationType,
          version,
          allRoutes: Array.from(allRoutes)
        });
      }
    }, [navigationType, stableLocationParam]);
    return Routes;
  };
  return (routes, locationArg) => {
    return /* @__PURE__ */ React.createElement(SentryRoutes, { routes, locationArg });
  };
}
function wrapPatchRoutesOnNavigation(opts, isMemoryRouter = false, capturedSpan) {
  if (!opts || !("patchRoutesOnNavigation" in opts) || typeof opts.patchRoutesOnNavigation !== "function") {
    return opts || {};
  }
  const originalPatchRoutes = opts.patchRoutesOnNavigation;
  return {
    ...opts,
    patchRoutesOnNavigation: async (args) => {
      const targetPath = args?.path;
      const activeRootSpan = getActiveRootSpan() ?? capturedSpan;
      if (!isMemoryRouter) {
        const originalPatch = args?.patch;
        const matches = args?.matches;
        if (originalPatch) {
          args.patch = (routeId, children) => {
            addRoutesToAllRoutes(children);
            if (matches && matches.length > 0) {
              const leafMatch = matches[matches.length - 1];
              const leafRoute = leafMatch?.route;
              if (leafRoute) {
                const matchingRoute = Array.from(allRoutes).find((route) => {
                  const idMatches = route.id !== void 0 && route.id === routeId;
                  const referenceMatches = route === leafRoute;
                  const pathMatches = route.path !== void 0 && leafRoute.path !== void 0 && route.path === leafRoute.path;
                  return idMatches || referenceMatches || pathMatches;
                });
                if (matchingRoute) {
                  addResolvedRoutesToParent(children, matchingRoute);
                }
              }
            }
            const spanJson = activeRootSpan ? spanToJSON(activeRootSpan) : void 0;
            if (targetPath && activeRootSpan && spanJson && !spanJson.timestamp && // Span hasn't ended yet
            spanJson.op === "navigation") {
              updateNavigationSpan(
                activeRootSpan,
                { pathname: targetPath, search: "", hash: "", state: null, key: "default" },
                Array.from(allRoutes),
                true,
                _matchRoutes
              );
            }
            return originalPatch(routeId, children);
          };
        }
      }
      const lazyLoadPromise = (async () => {
        const contextToken = setNavigationContext(targetPath, activeRootSpan);
        let result;
        try {
          result = await originalPatchRoutes(args);
        } finally {
          clearNavigationContext(contextToken);
          if (activeRootSpan) {
            resolveDeferredLazyRoutePromise(activeRootSpan);
          }
        }
        const spanJson = activeRootSpan ? spanToJSON(activeRootSpan) : void 0;
        if (activeRootSpan && spanJson && !spanJson.timestamp && // Span hasn't ended yet
        spanJson.op === "navigation") {
          const pathname = targetPath;
          if (pathname) {
            updateNavigationSpan(
              activeRootSpan,
              { pathname, search: "", hash: "", state: null, key: "default" },
              Array.from(allRoutes),
              false,
              _matchRoutes
            );
          }
        }
        return result;
      })();
      if (activeRootSpan) {
        trackLazyRouteLoad(activeRootSpan, lazyLoadPromise);
      }
      return lazyLoadPromise;
    }
  };
}
function handleNavigation(opts) {
  const { location, routes, navigationType, version, matches, basename, allRoutes: allRoutes2 } = opts;
  const branches = Array.isArray(matches) ? matches : _matchRoutes(allRoutes2 || routes, location, basename);
  const client = getClient();
  if (!client || !CLIENTS_WITH_INSTRUMENT_NAVIGATION.has(client)) {
    return;
  }
  const activeRootSpan = getActiveRootSpan();
  if (activeRootSpan && spanToJSON(activeRootSpan).op === "pageload" && navigationType === "POP") {
    return;
  }
  if ((navigationType === "PUSH" || navigationType === "POP") && branches) {
    const [name, source] = resolveRouteNameAndSource(
      location,
      allRoutes2 || routes,
      allRoutes2 || routes,
      branches,
      basename,
      _lazyRouteManifest,
      _enableAsyncRouteHandlers
    );
    const locationKey = computeLocationKey(location);
    const trackedNav = activeNavigationSpans.get(client);
    const trackedSpanHasEnded = trackedNav && !trackedNav.isPlaceholder ? !!spanToJSON(trackedNav.span).timestamp : false;
    const { skip, shouldUpdate } = shouldSkipNavigation(trackedNav, locationKey, name, trackedSpanHasEnded);
    if (skip) {
      if (shouldUpdate && trackedNav) {
        const oldName = trackedNav.routeName;
        if (trackedNav.isPlaceholder) {
          trackedNav.routeName = name;
          DEBUG_BUILD && debug.log(
            `[Tracing] Updated placeholder navigation name from "${oldName}" to "${name}" (will apply to real span)`
          );
        } else {
          trackedNav.span.updateName(name);
          trackedNav.span.setAttribute(SEMANTIC_ATTRIBUTE_SENTRY_SOURCE, source);
          if (source === "route") {
            trackedNav.span.setAttribute(URL_TEMPLATE, name);
          }
          addNonEnumerableProperty(
            trackedNav.span,
            "__sentry_navigation_name_set__",
            true
          );
          trackedNav.routeName = name;
          DEBUG_BUILD && debug.log(`[Tracing] Updated navigation span name from "${oldName}" to "${name}"`);
        }
      } else {
        DEBUG_BUILD && debug.log(`[Tracing] Skipping duplicate navigation for location: ${locationKey}`);
      }
      return;
    }
    const placeholderSpan = { end: () => {
    } };
    const placeholderEntry = {
      span: placeholderSpan,
      routeName: name,
      pathname: location.pathname,
      locationKey,
      isPlaceholder: true
    };
    activeNavigationSpans.set(client, placeholderEntry);
    let navigationSpan;
    try {
      navigationSpan = startBrowserTracingNavigationSpan(client, {
        name: placeholderEntry.routeName,
        // Use placeholder's routeName in case it was updated
        attributes: {
          [SEMANTIC_ATTRIBUTE_SENTRY_SOURCE]: source,
          [SEMANTIC_ATTRIBUTE_SENTRY_OP]: "navigation",
          [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: `auto.navigation.react.reactrouter${version ? `_v${version}` : ""}`,
          ...source === "route" && { [URL_TEMPLATE]: placeholderEntry.routeName }
        }
      });
    } catch (e) {
      activeNavigationSpans.delete(client);
      throw e;
    }
    if (navigationSpan) {
      activeNavigationSpans.set(client, {
        span: navigationSpan,
        routeName: placeholderEntry.routeName,
        // Use the (potentially updated) placeholder routeName
        pathname: location.pathname,
        locationKey
      });
      patchSpanEnd(navigationSpan, location, routes, basename, "navigation");
    } else {
      activeNavigationSpans.delete(client);
    }
  }
}
function addRoutesToAllRoutes(routes) {
  routes.forEach((route) => {
    const extractedChildRoutes = getChildRoutesRecursively(route);
    extractedChildRoutes.forEach((r) => {
      allRoutes.add(r);
    });
  });
}
function getChildRoutesRecursively(route, allRoutes2 = /* @__PURE__ */ new Set()) {
  if (!allRoutes2.has(route)) {
    allRoutes2.add(route);
    if (route.children && !route.index) {
      route.children.forEach((child) => {
        const childRoutes = getChildRoutesRecursively(child, allRoutes2);
        childRoutes.forEach((r) => {
          allRoutes2.add(r);
        });
      });
    }
  }
  return allRoutes2;
}
function updatePageloadTransaction({
  activeRootSpan,
  location,
  routes,
  matches,
  basename,
  allRoutes: allRoutes2
}) {
  const branches = Array.isArray(matches) ? matches : _matchRoutes(allRoutes2 || routes, location, basename);
  if (branches) {
    const [name, source] = resolveRouteNameAndSource(
      location,
      allRoutes2 || routes,
      allRoutes2 || routes,
      branches,
      basename,
      _lazyRouteManifest,
      _enableAsyncRouteHandlers
    );
    getCurrentScope().setTransactionName(name || "/");
    if (activeRootSpan) {
      activeRootSpan.updateName(name);
      activeRootSpan.setAttribute(SEMANTIC_ATTRIBUTE_SENTRY_SOURCE, source);
      if (source === "route") {
        activeRootSpan.setAttribute(URL_TEMPLATE, name);
      }
      patchSpanEnd(activeRootSpan, location, routes, basename, "pageload");
    }
  } else if (activeRootSpan) {
    patchSpanEnd(activeRootSpan, location, routes, basename, "pageload");
  }
}
function shouldUpdateWildcardSpanName(currentName, currentSource, newName, newSource, allowNoCurrentName = false) {
  if (!newName) {
    return false;
  }
  if (!currentName && allowNoCurrentName) {
    return true;
  }
  const hasWildcard = currentName && transactionNameHasWildcard(currentName);
  if (hasWildcard && newSource === "route" && !transactionNameHasWildcard(newName)) {
    return true;
  }
  if (currentSource !== "route" && newSource === "route") {
    return true;
  }
  return false;
}
function tryUpdateSpanNameBeforeEnd(span, spanJson, currentName, location, routes, basename, spanType, allRoutes2) {
  try {
    const currentSource = spanJson.data?.[SEMANTIC_ATTRIBUTE_SENTRY_SOURCE];
    if (currentSource === "route" && currentName && !transactionNameHasWildcard(currentName)) {
      return;
    }
    const currentAllRoutes = Array.from(allRoutes2);
    const routesToUse = currentAllRoutes.length > 0 ? currentAllRoutes : routes;
    const branches = _matchRoutes(routesToUse, location, basename);
    if (!branches) {
      return;
    }
    const [name, source] = resolveRouteNameAndSource(
      location,
      routesToUse,
      routesToUse,
      branches,
      basename,
      _lazyRouteManifest,
      _enableAsyncRouteHandlers
    );
    const isImprovement = shouldUpdateWildcardSpanName(currentName, currentSource, name, source, true);
    const spanNotEnded = spanType === "pageload" || !spanJson.timestamp;
    if (isImprovement && spanNotEnded) {
      span.updateName(name);
      span.setAttribute(SEMANTIC_ATTRIBUTE_SENTRY_SOURCE, source);
      if (source === "route") {
        span.setAttribute(URL_TEMPLATE, name);
      }
    }
  } catch (error) {
    DEBUG_BUILD && debug.warn(`Error updating span details before ending: ${error}`);
  }
}
function patchSpanEnd(span, location, routes, basename, spanType) {
  const patchedPropertyName = `__sentry_${spanType}_end_patched__`;
  const hasEndBeenPatched = span?.[patchedPropertyName];
  if (hasEndBeenPatched || !span.end) {
    return;
  }
  const originalEnd = span.end.bind(span);
  let endCalled = false;
  span.end = function patchedEnd(...args) {
    if (endCalled) {
      return;
    }
    endCalled = true;
    const endTimestamp = args.length > 0 ? args[0] : Date.now() / 1e3;
    const spanJson = spanToJSON(span);
    const currentName = spanJson.description;
    const currentSource = spanJson.data?.[SEMANTIC_ATTRIBUTE_SENTRY_SOURCE];
    const cleanupNavigationSpan = () => {
      const client = getClient();
      if (client && spanType === "navigation") {
        const trackedNav = activeNavigationSpans.get(client);
        if (trackedNav && trackedNav.span === span) {
          activeNavigationSpans.delete(client);
        }
      }
    };
    const pendingPromises = pendingLazyRouteLoads.get(span);
    const mayHaveLazyRoutes = span.__sentry_may_have_lazy_routes__;
    const hasPendingOrMayHaveLazyRoutes = pendingPromises && pendingPromises.size > 0 || mayHaveLazyRoutes;
    const shouldWaitForLazyRoutes = hasPendingOrMayHaveLazyRoutes && currentName && (transactionNameHasWildcard(currentName) || currentSource !== "route");
    if (shouldWaitForLazyRoutes) {
      if (_lazyRouteTimeout === 0) {
        tryUpdateSpanNameBeforeEnd(span, spanJson, currentName, location, routes, basename, spanType, allRoutes);
        cleanupNavigationSpan();
        originalEnd(endTimestamp);
        return;
      }
      const timeoutPromise = new Promise((r) => setTimeout(r, _lazyRouteTimeout));
      let waitPromise;
      if (pendingPromises && pendingPromises.size > 0) {
        const allSettled = Promise.allSettled(pendingPromises).then(() => {
        });
        waitPromise = _lazyRouteTimeout === Infinity ? allSettled : Promise.race([allSettled, timeoutPromise]);
      } else {
        waitPromise = timeoutPromise;
      }
      waitPromise.then(() => {
        const updatedSpanJson = spanToJSON(span);
        tryUpdateSpanNameBeforeEnd(
          span,
          updatedSpanJson,
          updatedSpanJson.description,
          location,
          routes,
          basename,
          spanType,
          allRoutes
        );
        cleanupNavigationSpan();
        originalEnd(endTimestamp);
      }).catch(() => {
        cleanupNavigationSpan();
        originalEnd(endTimestamp);
      });
      return;
    }
    tryUpdateSpanNameBeforeEnd(span, spanJson, currentName, location, routes, basename, spanType, allRoutes);
    cleanupNavigationSpan();
    originalEnd(endTimestamp);
  };
  addNonEnumerableProperty(span, patchedPropertyName, true);
}
function createV6CompatibleWithSentryReactRouterRouting(Routes, version) {
  if (!_useEffect || !_useLocation || !_useNavigationType || !_createRoutesFromChildren || !_matchRoutes) {
    DEBUG_BUILD && debug.warn(`reactRouterV6Instrumentation was unable to wrap Routes because of one or more missing parameters.
      useEffect: ${_useEffect}. useLocation: ${_useLocation}. useNavigationType: ${_useNavigationType}.
      createRoutesFromChildren: ${_createRoutesFromChildren}. matchRoutes: ${_matchRoutes}.`);
    return Routes;
  }
  const SentryRoutes = (props) => {
    const isMountRenderPass = React.useRef(true);
    const location = _useLocation();
    const navigationType = _useNavigationType();
    useIsomorphicLayoutEffect(
      () => {
        const routes = _createRoutesFromChildren(props.children);
        if (isMountRenderPass.current) {
          addRoutesToAllRoutes(routes);
          updatePageloadTransaction({
            activeRootSpan: getActiveRootSpan(),
            location,
            routes,
            allRoutes: Array.from(allRoutes)
          });
          isMountRenderPass.current = false;
        } else {
          handleNavigation({ location, routes, navigationType, version, allRoutes: Array.from(allRoutes) });
        }
      },
      // Re-run only on location/navigation changes, not children changes
      [location, navigationType]
    );
    return /* @__PURE__ */ React.createElement(Routes, { ...props });
  };
  hoistNonReactStatics(SentryRoutes, Routes);
  return SentryRoutes;
}

export { addResolvedRoutesToParent, addRoutesToAllRoutes, allRoutes, computeLocationKey, createReactRouterV6CompatibleTracingIntegration, createV6CompatibleWithSentryReactRouterRouting, createV6CompatibleWrapCreateBrowserRouter, createV6CompatibleWrapCreateMemoryRouter, createV6CompatibleWrapUseRoutes, handleNavigation, processResolvedRoutes, shouldSkipNavigation, updateNavigationSpan };
